# Portfolio Project: Online Retail EDA with Python

## Project Overview  

This project explores an **online retail transactions dataset**, focusing on **data cleaning, exploratory data analysis (EDA), and deriving business insights**. The dataset contains information about customer purchases, including invoice details, product descriptions, quantities, prices, and customer IDs. 


### Objectives  
- Perform **Exploratory Data Analysis (EDA)** to identify key trends.  
- Analyze **sales performance, customer behavior, and popular products**.  
- Provide **data-driven recommendations** to optimize online retail strategies.  


### My Approach  
To tackle this project, I’ll start by **ETL (Extract, Transform, Load)** to clean and prepare the dataset. Then, I’ll conduct in-depth analysis to identify key trends and insights like **busiest sales periods, top-selling products, and high-value customers**. Let's dive in!  


## Dataset Overview


For this project, I'll be working with the **Online Retail** dataset, which contains transactional data from an online store between 2010 and 2011. The dataset is in a `.csv` file named **`online_retail.csv`**, and it includes details about purchases such as product descriptions, quantities, prices, timestamps, and customer IDs.  


### Data Columns  
The dataset consists of the following fields:  
- **InvoiceNo** – Unique invoice number for each transaction.  
- **StockCode** – Unique product identifier.  
- **Description** – Product name/description.  
- **Quantity** – Number of units purchased.  
- **InvoiceDate** – Timestamp of the transaction.  
- **UnitPrice** – Price per unit of the product.  
- **CustomerID** – Unique identifier for each customer.  
- **Country** – Country where the transaction took place.  


Let's dive in and explore the dataset!  

### My Approach

To analyze this dataset effectively and derive meaningful insights, the workflow is divided into the following key stages:

1. **Load the Data**  
   Import the dataset into a Pandas DataFrame and inspect its structure and sample records.

2. **Data Cleaning**  
   Handle missing values, inconsistent entries, and remove irrelevant data to ensure accuracy.

3. **Basic Statistics**  
   Explore summary statistics to understand distributions and spot potential issues early on.

4. **Feature Engineering**  
   Create new variables such as `Revenue`, `InvoiceMonth`, and date-based features to enhance analysis.

5. **Exploratory Data Analysis (EDA)**  
   Investigate sales trends, customer behavior, product popularity, and country-wise performance.

6. **Outlier Detection**  
   Identify and manage anomalous transactions that could skew the results.

7. **Visualizations**  
   Use charts (e.g., histograms, time series, bar plots) to better communicate patterns and findings.

8. **Insights and Recommendations**  
   Summarize key takeaways and provide actionable recommendations based on the analysis.

Let’s dive in and explore the dataset!



# ETL

## 1. Load the data

Import the required libraries and load the dataset.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("source/online_retail.csv", encoding="ISO-8859-1")  # We use encoding to avoid UnicodeDecodeError (or encoding="Windows-1252")

Explore and familiarize with the dataset

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [3]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

## 2. Data Cleaning

Now that we have identified the data types of each column and detected any missing (null) values, we have a clearer understanding of how to approach the ETL process.

Before proceeding, let's create a copy of the dataframe to preserve the original data in its unaltered state.

In [4]:
df_clean = df.copy()

### Type Casting

With the copy created, we will begin by modifying the data types of specific columns.  
In this case, we will convert the `Country`, `InvoiceNo`, and `StockCode` columns from the object type to the category type.  
This transformation will optimize memory usage and improve performance when handling these columns in Pandas.


In [5]:
df_clean['Country'] = df_clean['Country'].astype('category')
df_clean['InvoiceNo'] = df_clean['InvoiceNo'].astype('category')
df_clean['StockCode'] = df_clean['StockCode'].astype('category')

# Ensure the data types where set correctly with: df_clean.info()

After that, we can also make sure that CustomerID is `int` instead of `float` to help Pandas process the information more efficiently

In [6]:
df_clean['CustomerID'] = df_clean['CustomerID'].astype('Int64')

Finally, let's turn `InvoiceDate` to **datetime**

In [7]:
df_clean['InvoiceDate'] = pd.to_datetime(df_clean['InvoiceDate'])

### Handling Missing Values

#### Handling missing `CustomerID` values

The dataset contains missing values in the `CustomerID` column, but these transactions are still valid purchases. Instead of dropping them or imputing arbitrary values (which could introduce bias), I will leave them as `NaN`.

 Why?
- Removing these rows would result in **loss of actual transaction data**.
- Imputing fake IDs would be **misleading**, as customer IDs are unique identifiers.
- Pandas and Matplotlib **handle NaN values gracefully** in most operations.

#### Handling missing `Description` values

The dataset contains null values in the `Description` column. Since these rows cannot be dropped without losing valuable data, we impute the missing descriptions using the corresponding `StockCode` values (which are complete and unique).

For that purpose, we follow this steps:
1. **Create a mapping dictionary** where each `StockCode` points to its correct `Description` (using only rows with non-null descriptions)
2. **Fill null values** by matching each missing `Description` with its `StockCode`'s known description

**Key Note**: If a `StockCode` has no valid description in the dataset, its `NaN` values will remain.

In [8]:
# Step 1: Map StockCode to Description (drop duplicates to ensure 1:1 mapping)
stock_to_desc = df_clean.dropna(subset=['Description']).drop_duplicates('StockCode').set_index('StockCode')['Description']

# Step 2: Fill NaN Descriptions using the mapped StockCode values
df_clean['Description'] = df_clean['Description'].fillna(df_clean['StockCode'].map(stock_to_desc))

After handling the preliminary missing values in the `Description` column, it's important to verify if any null values still remain. We will perform this check to ensure that all missing descriptions have been properly handled before moving forward with further analysis.

To do so, we'll check for any remaining nulls in the column.

In [9]:
# This will give us an updated count of the missing values in the 'Description' column
print(f'Original Description column null values: {df['Description'].isna().sum()}')
print(f'Updated Description column null values: {df_clean['Description'].isna().sum()}')

Original Description column null values: 1454
Updated Description column null values: 112


##### Imputing Remaining Null Values

After checking for null values, we found that 112 missing descriptions remain out of the initial 1,454 null values. To ensure we don't lose valuable transaction data, we will impute these remaining null values with the placeholder `'Unknown'`. This decision allows us to retain all rows in the dataset while clearly marking the transactions with missing descriptions.

In [10]:
df_clean['Description'] = df_clean['Description'].fillna('Unknown')

# To make sure this worked as intended: print(df_clean['Description'].isnull().sum())

By doing this, we preserve the full dataset while handling missing descriptions in a way that keeps the integrity of our analysis intact.

### String Processing

Before removing duplicates, we need to ensure that all truly identical rows are recognized as such by Pandas. To achieve this, we will standardize string formatting to eliminate inconsistencies.

We will focus on two key columns: `Description` and `Country`, as they contain string-type data.

`Description`: We will remove leading, trailing, and extra in-between whitespaces and standardize all text to uppercase for consistency.

`Country`: Similarly, we will trim unnecessary spaces and format country names in title case (first letter uppercase, the rest lowercase).

These transformations will help ensure that duplicate records are correctly identified and handled in the next step.

In [11]:
# Clean Description: Remove leading/trailing spaces, handle in-between extra spaces, and standardize to lowercase
df_clean['Description'] = df_clean['Description'].str.strip().str.replace(r'\s+', ' ', regex=True).str.upper()

# Clean Country: Remove leading/trailing spaces, handle in-between extra spaces, and title-case the country names
df_clean['Country'] = df_clean['Country'].str.strip().str.replace(r'\s+', ' ', regex=True).str.title()

### Removing Duplicates

Duplicated values can introduce bias and lead to incorrect insights, making it essential to handle them properly.

To begin, we will check for duplicate records in the dataset. Since individual columns may contain duplicate values, our focus will be on identifying and removing rows where all columns are identical.

For this, we will use Pandas' `.drop_duplicates()` method, which efficiently eliminates fully duplicated rows, ensuring data integrity for analysis.

In [12]:
print(f'Number of duplicate rows: {df_clean.duplicated().sum()}')

Number of duplicate rows: 5268


In [13]:
# Remove exact duplicate rows
df_clean = df_clean.drop_duplicates()

### Validating Negative Values

Based on the results of `df.describe()` in the *Load the Data* step, we identified negative values in the `Quantity` and `UnitPrice` columns. Since these values are not expected in a standard sales dataset, we will handle them systematically.

In [14]:
# Check for negative values in Quantity and UnitPrice
print(df_clean[df_clean['Quantity'] < 0].shape[0])
print(df_clean[df_clean['UnitPrice'] < 0].shape[0])

10587
2


Since the number of negative values in the `Quantity` column is significantly higher, we will address them first.

#### Analizing the negative values in `Quantity`

In [15]:
# Take a look at the negative values in Quantity and look for patterns
df_clean[df_clean['Quantity'] < 0].sample(20)

InvoiceNo StockCode                       Description  Quantity  \
44205    C540158     22968         ROSE COTTAGE KEEPSAKE BOX        -2   
188311   C553032     22148             EASTER CRAFT 4 CHICKS       -12   
393997   C570867     23256        CHILDRENS CUTLERY SPACEBOY        -4   
366477    568724     84816                           DAMAGES       -21   
70651    C542089     22300      COFFEE MUG DOG + BALL DESIGN        -2   
280898   C561510     22729       ALARM CLOCK BAKELIKE ORANGE        -1   
127557   C547232     85053        FRENCH ENAMEL CANDLEHOLDER        -1   
491733   C578073         M                            MANUAL       -36   
244401   C558522     22429        ENAMEL MEASURING JUG CREAM        -5   
385925   C570221     22470             HEART OF WICKER LARGE       -33   
225294   C556647     21870  I CAN ONLY PLEASE ONE PERSON MUG       -12   
225430   C556673     22656      VINTAGE BLUE KITCHEN CABINET        -1   
1986     C536548     22580      ADVENT CALENDAR GINGHAM SACK        -4   
293883   C562651     21991   BOHEMIAN COLLAGE STATIONERY SET       -12   
100710   C544831     22974          CHILDRENS DOLLY GIRL MUG        -2   
422617   C573098     23092   LARGE ANTIQUE WHITE PHOTO FRAME        -1   
340389   C566626     22461              SAVOY ART DECO CLOCK        -2   
537572   C581316     23174          REGENCY SUGAR BOWL GREEN        -1   
270323   C560561     22171  3 HOOK PHOTO SHELF ANTIQUE WHITE        -1   
215597   C555723     22171  3 HOOK PHOTO SHELF ANTIQUE WHITE        -2   

               InvoiceDate  UnitPrice  CustomerID         Country  
44205  2011-01-05 11:42:00       9.95       12471         Germany  
188311 2011-05-12 19:44:00       1.95       13320  United Kingdom  
393997 2011-10-12 16:17:00       4.15       12607             Usa  
366477 2011-09-28 16:27:00       0.00        <NA>  United Kingdom  
70651  2011-01-25 12:39:00       2.55       12681          France  
280898 2011-07-27 14:51:00       3.75       14534  United Kingdom  
127557 2011-03-21 16:15:00       2.10       12994  United Kingdom  
491733 2011-11-22 16:02:00       0.56       18139  United Kingdom  
244401 2011-06-30 10:33:00       4.25       15932  United Kingdom  
385925 2011-10-09 12:56:00       2.55       13113  United Kingdom  
225294 2011-06-13 16:14:00       1.25       13012  United Kingdom  
225430 2011-06-13 17:08:00     125.00       16949  United Kingdom  
1986   2010-12-01 14:33:00       5.95       12472         Germany  
293883 2011-08-08 13:25:00       1.25       16951  United Kingdom  
100710 2011-02-24 10:01:00       1.65       15146  United Kingdom  
422617 2011-10-27 14:45:00       7.90       15088  United Kingdom  
340389 2011-09-14 10:07:00       3.95       15380  United Kingdom  
537572 2011-12-08 11:46:00       4.15       12523          France  
270323 2011-07-19 13:28:00       8.50       17651  United Kingdom  
215597 2011-06-06 16:21:00       8.50       15737  United Kingdom

Upon examining the data, we can observe some patterns:  

1. Most `InvoiceNo` values associated with negative quantities begin with "C," which likely indicates a **credit transaction** (returns).  
2. Some `Description` values suggest special cases, such as "DAMAGED," "DISCOUNT," or "?".  

To better understand these cases, we will analyze how many unique `Description` values are associated with negative quantities.


In [16]:
negative_descriptions = df_clean[df_clean['Quantity'] < 0]['Description'].value_counts()
print(negative_descriptions.count())

2471


Given the large number of unique descriptions associated with negative quantities, we will focus on the most frequently occurring ones. This will help us identify common patterns and determine which descriptions may represent special cases that require specific handling.

In [17]:
print(negative_descriptions.head(30))  # Show the top 30 most frequent negative descriptions

Description
MANUAL                                 244
REGENCY CAKESTAND 3 TIER               180
POSTAGE                                126
CHECK                                  123
UNKNOWN                                 97
JAM MAKING SET WITH JARS                87
DISCOUNT                                77
SET OF 3 CAKE TINS PANTRY DESIGN        75
SAMPLES                                 61
DAMAGED                                 57
STRAWBERRY CERAMIC TRINKET BOX          54
ROSES REGENCY TEACUP AND SAUCER         54
RECIPE BOX PANTRY YELLOW DESIGN         47
DAMAGES                                 46
JUMBO BAG RED RETROSPOT                 44
LUNCH BAG RED RETROSPOT                 44
WOOD 2 DRAWER CABINET WHITE FINISH      43
RED RETROSPOT CAKE STAND                42
WHITE HANGING HEART T-LIGHT HOLDER      42
GREEN REGENCY TEACUP AND SAUCER         42
?                                       42
SMALL GLASS HEART TRINKET POT           40
SET OF 3 REGENCY CAKE TINS              37

#### Handling the negative values in `Quantity`

The output shows that some `Description` values represent regular products, while others indicate special cases (discounts, damaged products, or ambiguous values like "?"). We implement a unified classification system:

1. Categorize All Transactions

    Create a `TransactionType` column with three distinct labels:

    * **Return:** Transactions where `InvoiceNo` starts with "C" (credit notes).
    * **SpecialCase:** Transactions with descriptions matching predefined non-product terms (DISCOUNT, DAMAGED, SAMPLES, ?, etc.).
    * **Sale:** All other regular transactions.

2. Process Negative Quantities

    * **Returns:** Keep negatives (valid refund records).
    * **Special Cases:** Preserve original values (context-dependent).
    * **Sales:** Convert negatives to positives (assumed data entry errors).

**Note:** To distinguish legitimate negative quantities from data entry errors:

* Identified the top 30 most frequent descriptions for negative quantities

* Manually selected non-product terms (e.g., `DISCOUNT`, `DAMAGED`, `?`)

Resulting in the curated `special_case_list` used for classification.

In [18]:
# Predefined list of special cases for descriptions
special_case_list = [
    'DISCOUNT', 'DAMAGED', 'DAMAGES', 'SAMPLES', 'CHECK', 'MANUAL', 'POSTAGE', 
    'UNKNOWN', '?', 'AMAZON FEE', 'DOTCOM POSTAGE'
]

In [19]:
def classify_transaction(invoice_no, description):
    """Classify transaction as 'Return', 'SpecialCase', or 'Sale'."""
    if str(invoice_no).startswith('C'):
        return 'Return'
    elif description in special_case_list:
        return 'SpecialCase'
    else:
        return 'Sale'

# Use .map() efficiently by applying it on a tuple of (InvoiceNo, Description)
df_clean['TransactionType'] = list(map(classify_transaction, df_clean['InvoiceNo'], df_clean['Description']))

In [20]:
# Convert negatives to positives ONLY for regular Sales
df_clean.loc[
    (df_clean['TransactionType'] == 'Sale') & 
    (df_clean['Quantity'] < 0), 
    'Quantity'
] = df_clean['Quantity'].abs()  # Or: *= -1

To verify this approach worked as intended, we check if there are any negative `Quantity` values remaining that are neither classified as "Return" nor "SpecialCase".

In [21]:
print(f"Non Return or SpecialCase negative `Quantity` values: {df_clean[
    (df_clean['Quantity'] < 0) & 
    (~df_clean['TransactionType'].isin(['Return', 'SpecialCase']))
].shape[0]}")

Non Return or SpecialCase negative `Quantity` values: 0


Since the result is 0, it confirms that all negative `Quantity` values have been correctly handled according to our classification. We can now proceed to the next step.

#### Analizing and handling the negative values in `UnitPrice`

As observed earlier, only two transactions have negative `UnitPrice` values. Given their small number, we can inspect them directly as follows:

In [22]:
df_clean[df_clean['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  ADJUST BAD DEBT         1 2011-08-12 14:51:00   
299984   A563187         B  ADJUST BAD DEBT         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country TransactionType  
299983  -11062.06        <NA>  United Kingdom            Sale  
299984  -11062.06        <NA>  United Kingdom            Sale

Since they are labeled as "ADJUST BAD DEBT," these transactions appear to represent financial adjustments rather than product sales.

As these records are valid, we will keep them in the dataset and classify them under "SpecialCase" in the `TransactionType` column while preserving their negative values.


In [23]:
# Ensure "ADJUST BAD DEBT" transactions are marked as SpecialCase
df_clean.loc[df_clean['Description'] == 'ADJUST BAD DEBT', 'TransactionType'] = 'SpecialCase'

### Handling Outliers

The output from our initial exploration of `UnitPrice` shows that some transactions contain exceptionally high values. Upon further inspection, we found that several of these do not represent actual product sales but rather special cases, such as fees, postage, or adjustments.

1. Identifying High `UnitPrice` Transactions
To detect potential outliers, we examined the top transactions sorted by `UnitPrice`:

- Initially, we used `df_clean.nlargest(10, 'UnitPrice')` to inspect the highest values.

- Many high `UnitPrice` values corresponded to non-product transactions, which are **not currently classified as SpecialCases** (e.g., "DOTCOM POSTAGE", "THROW AWAY", "MOULDY, THROWN AWAY.").

- To refine our analysis, we iteratively excluded known non-product descriptions and re-ran the analysis to identify remaining cases.

In [24]:
df_clean.nlargest(10, 'UnitPrice')

InvoiceNo  StockCode      Description  Quantity         InvoiceDate  \
222681   C556445          M           MANUAL        -1 2011-06-10 15:31:00   
524602   C580605  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:36:00   
43702    C540117  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:55:00   
43703    C540118  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:57:00   
15016    C537630  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:04:00   
15017     537632  AMAZONFEE       AMAZON FEE         1 2010-12-07 15:08:00   
16356    C537651  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:49:00   
16232    C537644  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:34:00   
524601   C580604  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:35:00   
299982   A563185          B  ADJUST BAD DEBT         1 2011-08-12 14:50:00   

        UnitPrice  CustomerID         Country TransactionType  
222681   38970.00       15098  United Kingdom          Return  
524602   17836.46        <NA>  United Kingdom          Return  
43702    16888.02        <NA>  United Kingdom          Return  
43703    16453.71        <NA>  United Kingdom          Return  
15016    13541.33        <NA>  United Kingdom          Return  
15017    13541.33        <NA>  United Kingdom     SpecialCase  
16356    13541.33        <NA>  United Kingdom          Return  
16232    13474.79        <NA>  United Kingdom          Return  
524601   11586.50        <NA>  United Kingdom          Return  
299982   11062.06        <NA>  United Kingdom     SpecialCase

2. Filtering Out Non-Product Cases
We applied a filtering step to exclude descriptions that represent fees, adjustments, or other special cases, ensuring that we focus on actual product-related outliers. The filtered descriptions include:

    "AMAZON FEE", "MANUAL", "DOTCOM POSTAGE", "BANK CHARGES", "ADJUST BAD DEBT", "POSTAGE", "DISCOUNT", "CRUK COMMISSION"

In [25]:
df_clean[~df_clean['Description'].isin(
    ["AMAZON FEE", "MANUAL", 'DOTCOM POSTAGE', 'BANK CHARGES', 'ADJUST BAD DEBT', 'POSTAGE', 'DISCOUNT', 'CRUK COMMISSION' 
    ])].nlargest(10, 'UnitPrice')

# 'DOTCOM POSTAGE', 'THROW AWAY', 'UNSALEABLE, DESTROYED.', 'MOULDY, THROWN AWAY.'

InvoiceNo StockCode                     Description  Quantity  \
222680    556444     22502  PICNIC BASKET WICKER 60 PIECES        60   
222682    556446     22502  PICNIC BASKET WICKER 60 PIECES         1   
242589   C558359         S                         SAMPLES        -1   
4989      536835     22655     VINTAGE RED KITCHEN CABINET         1   
32484     539080     22655     VINTAGE RED KITCHEN CABINET         1   
36165    C539438     22655     VINTAGE RED KITCHEN CABINET        -1   
51636     540647     22655     VINTAGE RED KITCHEN CABINET         1   
82768     543253     22655     VINTAGE RED KITCHEN CABINET         1   
87141    C543632     22655     VINTAGE RED KITCHEN CABINET        -1   
118769    546480     22656    VINTAGE BLUE KITCHEN CABINET         1   

               InvoiceDate  UnitPrice  CustomerID         Country  \
222680 2011-06-10 15:28:00      649.5       15098  United Kingdom   
222682 2011-06-10 15:33:00      649.5       15098  United Kingdom   
242589 2011-06-28 15:10:00      570.0        <NA>  United Kingdom   
4989   2010-12-02 18:06:00      295.0       13145  United Kingdom   
32484  2010-12-16 08:41:00      295.0       16607  United Kingdom   
36165  2010-12-17 15:11:00      295.0       16607  United Kingdom   
51636  2011-01-10 14:57:00      295.0       17406  United Kingdom   
82768  2011-02-04 15:32:00      295.0       14842  United Kingdom   
87141  2011-02-10 16:22:00      295.0       14842  United Kingdom   
118769 2011-03-14 11:38:00      295.0       13452  United Kingdom   

       TransactionType  
222680            Sale  
222682            Sale  
242589          Return  
4989              Sale  
32484             Sale  
36165           Return  
51636             Sale  
82768             Sale  
87141           Return  
118769            Sale

To handle the identified outliers, we expand the `special_case_list` to include these descriptions and reuse the `classify_transaction` function, ensuring they are correctly categorized in the dataset.

In [26]:
# Expanding the special_case_list with newly identified non-product descriptions
special_case_list.extend([
    "DOTCOM POSTAGE", "THROW AWAY", "UNSALEABLE, DESTROYED.", "MOULDY, THROWN AWAY.",
    "AMAZON FEE", "MANUAL", "BANK CHARGES", "ADJUST BAD DEBT", "POSTAGE", "DISCOUNT", "CRUK COMMISSION"
])

# Reapplying the classification function to update TransactionType
df_clean['TransactionType'] = list(map(classify_transaction, df_clean['InvoiceNo'], df_clean['Description']))


Then, we can re-run the high `UnitPrice`analysis to confirm that all outliers are either properly classified or remain legitimate product sales.

By implementing this refined approach, we ensure that outlier removal is not arbitrary but instead data-driven, focusing on legitimate product transactions while flagging non-product entries as SpecialCases.

In [27]:
df_clean.nlargest(10, 'UnitPrice')

InvoiceNo  StockCode      Description  Quantity         InvoiceDate  \
222681   C556445          M           MANUAL        -1 2011-06-10 15:31:00   
524602   C580605  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:36:00   
43702    C540117  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:55:00   
43703    C540118  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:57:00   
15016    C537630  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:04:00   
15017     537632  AMAZONFEE       AMAZON FEE         1 2010-12-07 15:08:00   
16356    C537651  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:49:00   
16232    C537644  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:34:00   
524601   C580604  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:35:00   
299982   A563185          B  ADJUST BAD DEBT         1 2011-08-12 14:50:00   

        UnitPrice  CustomerID         Country TransactionType  
222681   38970.00       15098  United Kingdom          Return  
524602   17836.46        <NA>  United Kingdom          Return  
43702    16888.02        <NA>  United Kingdom          Return  
43703    16453.71        <NA>  United Kingdom          Return  
15016    13541.33        <NA>  United Kingdom          Return  
15017    13541.33        <NA>  United Kingdom     SpecialCase  
16356    13541.33        <NA>  United Kingdom          Return  
16232    13474.79        <NA>  United Kingdom          Return  
524601   11586.50        <NA>  United Kingdom          Return  
299982   11062.06        <NA>  United Kingdom     SpecialCase

## 3. Basic Statistics

After cleaning the dataset, we proceed to explore its basic statistics. This initial analysis provides a foundation for understanding the data’s structure, identifying early patterns, and evaluating potential anomalies or irregularities. It sets the stage for more advanced analytical steps by offering a general snapshot of the dataset’s composition.

We begin by generating summary statistics and exploring key aspects of the data.

In [28]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   InvoiceNo        536641 non-null  category      
 1   StockCode        536641 non-null  category      
 2   Description      536641 non-null  object        
 3   Quantity         536641 non-null  int64         
 4   InvoiceDate      536641 non-null  datetime64[ns]
 5   UnitPrice        536641 non-null  float64       
 6   CustomerID       401604 non-null  Int64         
 7   Country          536641 non-null  object        
 8   TransactionType  536641 non-null  object        
dtypes: Int64(1), category(2), datetime64[ns](1), float64(1), int64(1), object(3)
memory usage: 52.8+ MB


**Key Observations:**

- **Rows**: 536,641 transactions
- **Columns**: 9
- Missing Values: Most columns are complete, except for CustomerID, which is missing in approximately 25% of the records
- **Derived Column**: A new column, `TransactionType`, classifies each row as a `Sale`, `Return`, or `SpecialCase` based on invoice patterns and product descriptions

In [29]:
df_clean['TransactionType'].value_counts(normalize=True)

TransactionType
Sale           0.977879
Return         0.017239
SpecialCase    0.004882
Name: proportion, dtype: float64

The majority of transactions are regular **Sales**, followed by a smaller proportion of **Returns** and **Special Cases**, as expected based on our previous classification process.

### Overview of each transaction type

Having all transactions categorized under `TransactionType`, we can now analyze each group independently without conflating their behaviors. This allows for more accurate insights and better handling of specific cases.

To begin, we use the `.describe()` method to generate summary statistics for each transaction type.

In [30]:
df_clean[df_clean['TransactionType'] == 'Sale'].drop(columns=['CustomerID']).describe()

Quantity                    InvoiceDate      UnitPrice
count  524770.000000                         524770  524770.000000
mean       10.987383  2011-07-04 13:00:30.277493504       3.274727
min         1.000000            2010-12-01 08:26:00       0.000000
25%         1.000000            2011-03-28 11:51:00       1.250000
50%         4.000000            2011-07-19 17:17:00       2.080000
75%        12.000000            2011-10-19 10:59:00       4.130000
max     80995.000000            2011-12-09 12:50:00     649.500000
std       159.878853                            NaN       4.460465

In [31]:
df_clean[df_clean['TransactionType'] == 'Return'].drop(columns=['CustomerID']).describe()

Quantity                    InvoiceDate     UnitPrice
count   9251.000000                           9251   9251.000000
mean     -29.787050  2011-06-26 07:06:45.231866880     48.570430
min   -80995.000000            2010-12-01 09:41:00      0.010000
25%       -6.000000            2011-03-21 16:26:00      1.450000
50%       -2.000000            2011-07-08 13:04:00      2.950000
75%       -1.000000            2011-10-06 20:36:00      5.950000
max       -1.000000            2011-12-09 11:58:00  38970.000000
std     1147.997592                            NaN    667.926393

In [32]:
df_clean[df_clean['TransactionType'] == 'SpecialCase'].drop(columns=['CustomerID']).describe()

Quantity                    InvoiceDate     UnitPrice
count  2620.000000                           2620   2620.000000
mean      4.115649  2011-06-29 00:50:42.847328256    121.476855
min   -3000.000000            2010-12-01 08:45:00 -11062.060000
25%       1.000000            2011-03-25 14:14:15      2.950000
50%       1.000000            2011-07-05 17:01:00     18.000000
75%       2.000000            2011-10-13 16:19:45    118.342500
max    5368.000000            2011-12-09 12:16:00  13541.330000
std     232.965881                            NaN    580.267542

**Key Takeaways:**

- **Returns** correctly reflect negative quantities, consistent with refund transactions.
- **Special Cases** exhibit high variability in both `Quantity` and `UnitPrice`, supporting their exclusion from regular sales analysis.
- **Sales** now contain only positive quantities and show a reasonable distribution of unit prices, indicating reliable data for core business analysis.


## 4. Feature Engineering

In this section, we enrich our dataset by creating new, meaningful features that will support future exploratory analysis and help uncover deeper insights.

The main goals of this step are to:

- Create a Revenue column to measure transaction value.
- Extract time-based features from the invoice date to enable seasonality and behavior analysis.
- Engineer RFM (Recency, Frequency, Monetary) features to evaluate customer behavior and segment them based on purchasing patterns.

These engineered features will serve as a strong foundation for both exploratory data analysis (EDA) and potential modeling in future steps.

We begin by creating a `Revenue` column for each transaction by multiplying `Quantity` by `UnitPrice`. This serves as the foundation for later customer value analysis.

In [33]:
df_clean['Revenue'] = df_clean['Quantity'] * df_clean['UnitPrice']

To enable temporal analysis, we extract multiple date and time features from the `InvoiceDate` column. These will be useful for analyzing seasonality, trends, and customer behavior over time.

In [34]:
df_clean['InvoiceYear'] = df_clean['InvoiceDate'].dt.year
df_clean['InvoiceMonth'] = df_clean['InvoiceDate'].dt.month
df_clean['InvoiceDay'] = df_clean['InvoiceDate'].dt.day
df_clean['InvoiceHour'] = df_clean['InvoiceDate'].dt.hour
df_clean['InvoiceWeekday'] = df_clean['InvoiceDate'].dt.day_name()

### RFM analysis feature engineering

To enhance our understanding of customer behavior, we’ll perform a **RFM (Recency, Frequency, Monetary) analysis** — a well-established method for customer segmentation based on purchasing patterns.

Since RFM relies on the `CustomerID` column, and we intentionally retained null values in our main dataset (`df_clean`) to avoid bias in other analyses, we’ll work with a filtered copy. This new DataFrame, `notnull_customer_df`, includes only sales transactions with valid `CustomerID` values to ensure accurate segmentation.

Additionally, RFM analysis requires a reference point in time to calculate `Recency`. We'll define this reference date as one day after the most recent InvoiceDate in the dataset.

We filter the dataset to include only regular sales (`TransactionType == 'Sale'`) and transactions with known `CustomerID`. This ensures a reliable and unbiased base for calculating RFM metrics.

In [35]:
notnull_customer_df = df_clean[
    (df_clean['TransactionType'] == 'Sale') &
    (df_clean['CustomerID'].notnull())
].copy()

To compute the **Recency** metric (how recently each customer purchased), we define a `reference_date` as one day after the most recent invoice in the dataset.

In [36]:

#Select the reference date for RFM analysis
reference_date = notnull_customer_df['InvoiceDate'].max() + pd.Timedelta(days=1)

To calculate RFM metrics, we group transactions by CustomerID and compute the three core RFM metrics:

- **Recency**: Days since last purchase
- **Frequency**: Number of unique invoices (transactions)
- **Monetary**: Total revenue generated

In [37]:
# 1. Group by CustomerID and aggregate Recency, Frequency, Monetary
df_rfm = notnull_customer_df.groupby('CustomerID').agg({
    'InvoiceDate': lambda x: (reference_date - x.max()).days,            # Recency
    'InvoiceNo': 'nunique',                                              # Frequency
    'UnitPrice': 'sum'                                                   # Temporary, we'll correct this below
}).rename(columns={
    'InvoiceDate': 'Recency',
    'InvoiceNo': 'Frequency',
})

# Replace the placeholder UnitPrice sum with actual customer revenue using the Revenue column
df_rfm['Monetary'] =  notnull_customer_df.groupby('CustomerID')['Revenue'].sum()


Let's review summary statistics for the RFM features to understand their ranges, distributions, and central tendencies.

In [38]:
df_rfm.describe()

Recency    Frequency     UnitPrice       Monetary
count  4335.000000  4335.000000   4335.000000    4335.000000
mean     92.684890     4.246367    260.894753    2017.050437
std     100.172622     7.643880    697.201332    8919.387033
min       1.000000     1.000000      0.000000       0.000000
25%      18.000000     1.000000     48.850000     304.250000
50%      51.000000     2.000000    120.950000     663.610000
75%     143.000000     5.000000    277.860000    1631.475000
max     374.000000   206.000000  22776.010000  279138.020000